### PR TITLE
417 add ConvertForMetrics post transform

### LINKED ISSUE #417: add `ConvertForMetrics` post transform
**Is your feature request related to a problem? Please describe.**
Need a post transform to convert model output for ignite metrics.
For example:
add `sigmoid`, add `softmax`, change to `One-Hot` format, add `argmax`, `round value` to 0 or 1, etc.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #417 .

### Description
This PR implemented ConvertForMetrics transform.

### Status
**Ready**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### CODE REVIEW

File: monai/transforms/post/array.py [hunk 1]
```diff
@@ -46,3 +47,68 @@ def __call__(self, img, to_onehot=None, num_classes=None):
...
+        self.add_sigmoid = add_sigmoid
+        self.add_softmax = add_softmax
+        self.add_argmax = add_argmax
+        self.to_onehot_y_pred = to_onehot_y_pred
+        self.to_onehot_y = to_onehot_y
+        self.n_classes = n_classes
+        self.round_values = round_values
+        self.logit_thresh = logit_thresh
+
+    def __call__(self, y_pred, y=None):
```
Comment (wyli) [inline monai/transforms/post/array.py:92]: I think this module does too many things, could decompose it into:
- `AsDiscrete`, for `argmax`, `to_one_hot`, `thresholding`, (round_values is not very accurate here)
- `NonLinear`, for `sigmoid`,`softmax`, other layers from `torch.nn`'s non linear functions collection https://pytorch.org/docs/stable/nn.html#non-linear-activations-other

also optionally we could have some basic input validation, e.g. sigmoid=True and softmax=True are not compatible options

File: monai/transforms/post/array.py [hunk 2]
```diff
@@ -13,6 +13,7 @@
 https://github.com/Project-MONAI/MONAI/wiki/MONAI_Design
```
Comment (wyli) [inline monai/transforms/post/array.py:13]: maybe add a comment here that this postprocessing module assumes shape `[batch, channels, (optional spatial dimensions)]`

also, [`write_nifti`](https://github.com/Project-MONAI/MONAI/blob/master/monai/data/nifti_writer.py#L19) `write_png` are also postprocessings, shall we move them here? (perhaps need to think about this, not a high priority task for now)

File: monai/transforms/post/array.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/transforms/post/array.py:92]: updated in the latest patch.
Thanks.

File: monai/transforms/post/array.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/transforms/post/array.py:13]: updated in the latest patch.
`write_nifiti` can be also used by pre-processing transform, to save some transform result.
`save_nifiti` can be really post_transforms I feel.
You can submit an task issue for that I think.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Thanks very much for your review.
I updated the code according to your comments.
Could you please help review it again?
Thanks.

File: monai/transforms/post/array.py [hunk 3]
```diff
@@ -46,3 +48,72 @@ def __call__(self, img, to_onehot=None, num_classes=None):
...
+
+    def __init__(self, add_sigmoid=False, add_softmax=False, other=None):
+        self.add_sigmoid = add_sigmoid
+        self.add_softmax = add_softmax
+        self.other = other
+
+    def __call__(self, img, add_sigmoid=None, add_softmax=None, other=None):
+        if add_sigmoid is True and add_softmax is True:
+            raise ValueError("add_sigmoid=True and add_softmax=True are not compatible.")
+        if self.add_sigmoid if add_sigmoid is None else add_sigmoid:
```
Comment (wyli) [inline monai/transforms/post/array.py:73]: `self.add_sigmoid if add_sigmoid is None else add_sigmoid` could be simplified as
`add_sigmoid or self.add_sigmoid`

File: monai/transforms/post/array.py [hunk 4]
```diff
@@ -46,3 +48,72 @@ def __call__(self, img, to_onehot=None, num_classes=None):
...
+
+    Args:
+        add_sigmoid (bool): whether to add sigmoid function to model output before transform.
+        add_softmax (bool): whether to add softmax function to model output before transform.
+        other (Callable): callable function to execute other activation layers, for example:
+            `other = lambda x: torch.tanh(x)`
+
+    """
+
+    def __init__(self, add_sigmoid=False, add_softmax=False, other=None):
```
Comment (wyli) [inline monai/transforms/post/array.py:65]: could replace `add_sigmoid` with `sigmoid`, `add_softmax` with `softmax`, and `AddActivations` with `Activations`, because I feel it's a bit confusing given we have `AddChannel`

File: monai/transforms/post/array.py [hunk 5]
```diff
@@ -46,3 +48,72 @@ def __call__(self, img, to_onehot=None, num_classes=None):
...
+    def __call__(self, img, add_sigmoid=None, add_softmax=None, other=None):
+        if add_sigmoid is True and add_softmax is True:
+            raise ValueError("add_sigmoid=True and add_softmax=True are not compatible.")
+        if self.add_sigmoid if add_sigmoid is None else add_sigmoid:
+            img = torch.sigmoid(img)
+        if self.add_softmax if add_softmax is None else add_softmax:
+            img = torch.softmax(img, dim=1)
+        act_func = self.other if other is None else other
+        if act_func is not None:
+            if not isinstance(act_func, Callable):
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/post/array.py:79]: updated in the latest patch.
Thanks.

File: monai/transforms/post/array.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/transforms/post/array.py:65]: updated in the latest patch.
Thanks.

File: monai/transforms/post/array.py (hunk 3 above)
Comment (Nic-Ma) [reply monai/transforms/post/array.py:73]: updated in the latest patch.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Thanks for your review.
I updated the code according to your comments.
Could you please help review it again?
I will also unify the `do_sigmoid`, `add_sigmoid`, `do_softmax`, etc. in DiceLoss, MeanDice and AUC metrics to `sigmoid`, `softmax` in another PR later.
Thanks.

File: tests/test_as_discreted.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        "logit_thresh": 0.5,
+    },
+    {"pred": torch.tensor([[[[0.0, 1.0]], [[2.0, 3.0]]]])},
+    {"pred_discrete": torch.tensor([[[[0.0, 0.0]], [[1.0, 1.0]]]])},
+    (1, 2, 1, 2),
+]
+
+
+class TestAsDiscreted(unittest.TestCase):
+    @parameterized.expand([TEST_CASE_2])
```
Comment (wyli) [inline tests/test_as_discreted.py:70]: test cases not in use?

File: docs/source/transforms.rst [hunk 7]
```diff
@@ -584,12 +597,25 @@ Dictionary-based Transforms
     :members:
     :special-members: __call__
 
+`AddActivationsd`
+~~~~~~~~~~~~~~~~~
+.. autoclass:: AddActivationsd
```
Comment (wyli) [inline docs/source/transforms.rst:602]: now this needs to be changed as well

File: tests/test_as_discreted.py (hunk 6 above)
Comment (Nic-Ma) [reply tests/test_as_discreted.py]: updated in the latest patch.
Thanks.

File: docs/source/transforms.rst (hunk 7 above)
Comment (Nic-Ma) [reply docs/source/transforms.rst]: updated in the latest patch.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Thanks for your review, I updated the code.
Could you please help review it again?

**Comment by wyli** [comment]
> Hi @wyli ,
> 
> Thanks for your review, I updated the code.
> Could you please help review it again?

thanks, this pr introduces a new warning for documentation building
```
MONAI/monai/transforms/post/array.py:docstring of monai.transforms.AsDiscrete:3: WARNING: Unexpected indentation.
```

**Comment by Nic-Ma** [comment]
> > Hi @wyli ,
> > Thanks for your review, I updated the code.
> > Could you please help review it again?
> 
> thanks, this pr introduces a new warning for documentation building
> 
> ```
> MONAI/monai/transforms/post/array.py:docstring of monai.transforms.AsDiscrete:3: WARNING: Unexpected indentation.
> ```

Thanks for your reminder. Fixed in the latest patch.
